test/api/scenes.ts

```diff
@@ -0,0 +1,170 @@
+
+  let scenes = scenesService.getScenes();
+  let scenesNames = scenes.map(scene => scene.name);
+
+  t.deepEqual(scenesNames, ['Scene', 'Scene2']);
+
+  scenesService.removeScene(scenes[1].id);
+  scenes = scenesService.getScenes();
+  scenesNames = scenes.map(scene => scene.name);
+
+  t.deepEqual(scenesNames, ['Scene']);
+});
+
+
+test('Switching between scenes', async t => {
+  const client = await getClient();
+  const scenesService = client.getResource<IScenesServiceApi>('ScenesService');
+
+  const scene = scenesService.getSceneByName('Scene');
+  const scene2 = scenesService.createScene('Scene2');
+
+
+  t.is(scene.id, scenesService.activeSceneId);
+
+  scenesService.makeSceneActive(scene2.id);
+
+  t.is(scene2.id, scenesService.activeSceneId);
+
+  scene2.remove();
+
+  t.is(scene.id, scenesService.activeSceneId);
+
+});
+
+
+test('Creating, fetching and removing scene-items', async t => {
+  const client = await getClient();
+  const scenesService = client.getResource<IScenesServiceApi>('ScenesService');
+
+  const scene = scenesService.getSceneByName('Scene');
+  const image1 = scene.createAndAddSource('Image1', 'image_source');
+  const image2 = scene.createAndAddSource('Image2', 'image_source');
+  t.is(image1['name'], 'Image1');
+
+  let items = scene.getItems();
+  let itemsNames = items.map(item => item['name']);
+  t.deepEqual(itemsNames, ['Image2', 'Image1']);
+
+
+  scene.removeItem(image2.sceneItemId);
+  items = scene.getItems();
+  itemsNames = items.map(item => item['name']);
+  t.deepEqual(itemsNames, ['Image1']);
+
+});
+
+
+test('Scenes events', async t => {
+  const client = await getClient();
+  const scenesService = client.getResource<IScenesServiceApi>('ScenesService');
+
+  let lastEventData = null;
+  const onEventHandler = (event: any) => {
+    lastEventData = event;
+  };
+
+  scenesService.sceneSwitched.subscribe(onEventHandler);
+  scenesService.sceneAdded.subscribe(onEventHandler);
+  scenesService.sceneRemoved.subscribe(onEventHandler);
+  scenesService.itemAdded.subscribe(onEventHandler);
+  scenesService.itemRemoved.subscribe(onEventHandler);
+  scenesService.itemUpdated.subscribe(onEventHandler);
+
+  const scene2 = scenesService.createScene('Scene2');
+  await sleep(100);
+
+  t.is(lastEventData && lastEventData['name'], 'Scene2');
+
+  const scene3 = scenesService.createScene('Scene3');
+  await sleep(100);
+
+  scenesService.makeSceneActive(scene2.id);
+  await sleep(100);
+  t.is(lastEventData && lastEventData['name'], 'Scene2');
+
+  scene3.remove();
+  await sleep(100);
+  t.is(lastEventData && lastEventData['name'], 'Scene3');
+
+
+  const image = scene2.createAndAddSource('image', 'image_source');
+  await sleep(100);
+  t.is(lastEventData && lastEventData['sceneItemId'], image.sceneItemId);
+
+
+  image.setVisibility(false);
+  await sleep(100);
+  t.is(lastEventData && lastEventData['visible'], false);
+
+  lastEventData = null;
+  image.remove();
+  await sleep(100);
```

Comment:
When we assert that an element is present on the page, it will search for the element for up to 2 seconds (or whatever time we set) before giving up.  This is nice because if the rendering is happening quickly, the tests don't spend unnecessary amounts of time waiting, but will wait a long time if they need to.

I think that ideally we would do something similar for API event tests.  API events, just like rendered DOM nodes, are asynchronous results that we expect to appear within a reasonable timeframe.  It seems like we should build a promise-based assertion helper that allows you to expect an event, maybe with a custom matcher function provided, and it will resolve as soon as it receives the event, or reject and fail the test if it reaches the timeout.  I think this system would be much more reliable when the system is running slowly, and would run much faster when things are running well.  In its current state, I am a little worried that this test may be unreliable, especially on slow CI VMs.

Reply:
You're right. Was too rushing while implementing this.